### PR TITLE
Add CLI for route flow counter feature

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
HLD: https://github.com/Azure/SONiC/pull/908
Command reference PR: https://github.com/Azure/sonic-utilities/pull/2069

#### What I did

Add CLIs for route flow counter feature

#### How I did it

1. Add show command `show flowcnt-route config` and command group `show flowcnt-route stats`
2. Add config command group `config flowcnt-route pattern` 
3. Add clear command group `sonic-clear flowcnt-route` 

#### How to verify it

1. Full unit test cover
2. Manual test
3. sonic-mgmt test cases

#### New command output (if the output of a command-line utility has changed)

Enable/disable configuration:

```
counterpoll flowcnt-route  <enable | disable>
Example:
admin@sonic:~$ counterpoll flowcnt-route enable
```

Polling interval configuration:

```
counterpoll flowcnt-route interval <time_in_msec> // default - 1000ms
Example:
admin@sonic:~$ counterpoll flowcnt-route interval 2000
```

Show configuration:

```
counterpoll show
Example:
admin@sonic:~$ counterpoll show
Type                              Interval (in ms)          Status
--------------------------        ------------------        --------
FLOW_CNT_ROUTE_STAT               default(10000)            disable
```

Config route pattern:

```
config flowcnt-route pattern <add | remove> [--vrf <vrf>] [--max <route-max>] <prefix-pattern>    // configure route pattern
Example:
admin@sonic:~$ config flowcnt-route pattern add --vrf Vrf_1 --max 50 2.2.0.0/16
Route Pattern Flow Counter configuration is successful

admin@sonic:~$ config flowcnt-route pattern remove --vrf Vrf_1 2.2.0.0/16
Route Pattern Flow Counter configuration is successful
```

Show configuration:

```
show flowcnt-route config
Example:
admin@sonic:~$ show flowcnt-route config
Route pattern          VRF                 Max
-----------------------------------------------
3.3.0.0/16             default             50
```

Show counters value:

```
show flowcnt-route stats   // show statistics of all route flow counters
Example:
admin@sonic:~$ show flowcnt-route stats
Route pattern       VRF               Matched routes           Packets          Bytes
--------------------------------------------------------------------------------------
3.3.0.0/16          default           3.3.1.0/24               100              4543
                                      3.3.2.3/32               3443             929229
                                      3.3.0.0/16               0                0


show flowcnt-route stats pattern [<prefix-pattern> [ --vrf <vrf>] ]   // show statistics of all routes matching the configured route pattern
Example:
admin@sonic:~$ show flowcnt-route stats pattern 3.3.0.0/16
Route pattern       VRF               Matched routes           Packets          Bytes
--------------------------------------------------------------------------------------
3.3.0.0/16          default           3.3.1.0/24               100              4543
                                      3.3.2.3/32               3443             929229
                                      3.3.0.0/16               0                0

show flowcnt-route stats route [<prefix> [ --vrf <vrf>] ]             // show statistics of the specific route matching the configured route pattern
Example:
admin@sonic:~$ show flowcnt-route stats route 3.3.3.2/32 --vrf Vrf_1
Route                     VRF              Route Pattern           Packets          Bytes
-----------------------------------------------------------------------------------------
3.3.3.2/32                Vrf_1            3.3.0.0/16              100              4543
```

Clear counters:

```
sonic-clear flowcnt-route    // clear all route flow counters
Example:
admin@sonic:~$ sonic-clear flowcnt-route
Route Flow Counters were successfully cleared

sonic-clear flowcnt-route pattern  [<prefix-pattern> [ --vrf <vrf>] ]   // clear flow counters of all routes matching the configured route pattern
Example:
admin@sonic:~$ sonic-clear flowcnt-route pattern 3.3.0.0/16 --vrf Vrf_1
Flow Counters of all routes matching the configured route pattern were successfully cleared

sonic-clear flowcnt-route route [<prefix> [ --vrf <vrf>] ]  // clear flow counters of the specific route matching the configured prefix
Example:
admin@sonic:~$ sonic-clear flowcnt-route route 3.3.3.2/32 --vrf Vrf_1
Flow Counters of the specified route were successfully cleared
```

### CODE REVIEW

**Comment by dgsudharsan** [comment]
@Junchao-Mellanox Can you please update the description to include the output of the new show commands introduced? Please also update the command reference

File: tests/counterpoll_input/config_db.json [hunk 1]
```diff
@@ -1,2668 +1,2671 @@
 {
```
Comment (dgsudharsan) [inline tests/counterpoll_input/config_db.json]: I believe there are so much white space removal from this file which makes it harder to review the diff. Can you please make sure only functional changes are present and revert it as previous.
We can raise the indentation and whitespace removal as a separate PR if wanted.

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: scripts/flow_counters_stat [hunk 2]
```diff
@@ -150,7 +155,7 @@ class FlowCounterStats(object):
 
             full_table_id = RATES_TABLE_PREFIX + counter_oid
             counter_data =  self.db.get(self.db.COUNTERS_DB, full_table_id, PPS_FIELD)
-            values.append(STATUS_NA if counter_data is None else counter_data)
```
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply scripts/flow_counters_stat:153]: Two reasons：
1. If we do a clear when the value is "N/A", the value "N/A" will be saved to a file, and next time we will see an exception when doing diff
2. If couner_data is None, it means syncd has not updated statistic to COUNTERS DB yet. It is safe to assume the value is 0 here, and after some interval, the value will be updated with real value.

File: tests/mock_tables/state_db.json [hunk 3]
```diff
@@ -748,5 +748,8 @@
         "linkmgrd_switch_standby_end": "2021-May-13 10:01:15.696728",
         "xcvrd_switch_standby_end": "2021-May-13 10:01:15.696051",
         "xcvrd_switch_standby_start": "2021-May-13 10:01:15.690835"
+    },
+    "FLOW_COUNTER_CAPABILITY_TABLE|route": {
```
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply tests/mock_tables/state_db.json:752]: We don't support it for trap. This is a new suggestion provided by Guohan in route flow counter HLD review. We could add the same logic for trap counter in future.

File: tests/flow_counter_stats_test.py [hunk 4]
```diff
@@ -207,3 +404,537 @@ def test_clear(self):
...
+
+        mock_input.return_value = False
+        vrf = 'Vrf1'
+        result = runner.invoke(
+            config.config.commands["flowcnt-route"].commands["pattern"].commands["add"],
+            ['--max', '50', '--vrf', vrf, prefix], obj=db
+        )
+
+        assert result.exit_code == 0
+        assert prefix in table
```
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply tests/flow_counter_stats_test.py:457]: vrf shall not in table because "mock_input.return_value = False"

**Comment by Junchao-Mellanox** [comment]
> @Junchao-Mellanox Can you please update the description to include the output of the new show commands introduced? Please also update the command reference

Updated description. There is another PR which updated the command reference.

File: scripts/flow_counters_stat (hunk 2 above)
[the remark this answers is not in the document]
Comment (dgsudharsan) [reply scripts/flow_counters_stat:153]: Thanks for the explanation. If the route flow counters are not supported I believe there will be no output in the first place due to the check done. Am I correct?

**Review by dgsudharsan — APPROVED** [review]

**Comment by dgsudharsan** [comment]
> > @Junchao-Mellanox Can you please update the description to include the output of the new show commands introduced? Please also update the command reference
> 
> Updated description. There is another PR which updated the command reference.

Please add the link here.

File: scripts/flow_counters_stat (hunk 2 above)
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply scripts/flow_counters_stat:153]: For a platform that support the feature, there still could be timing issue. Consider following flow:
1. User enables route flow counters
2. Syncd has not updated the data to COUNTERS DB
3. User issue "show flowcnt-route stats"

**Comment by Junchao-Mellanox** [comment]
Did a rebase, hope it can fix the vstest issue.

**Review by dgsudharsan — DISMISSED** [review]

**Review by keboliu — DISMISSED** [review]

**Review by dgsudharsan — APPROVED** [review]

**Review by liat-grozovik — APPROVED** [review]

**Pull request merged**